src/Elastic.Apm/Cloud/AzureAppServiceMetadataProvider.cs

```diff
@@ -14,110 +13,40 @@ namespace Elastic.Apm.Cloud
 	/// <summary>
 	/// Provides cloud metadata for Microsoft Azure App Services
 	/// </summary>
-	public class AzureAppServiceMetadataProvider : ICloudMetadataProvider

```

Comment:
Hmm.. I also think this type should be `internal` and I don't know why this was `public` in the first place.

But making it `internal` from `public` is technically a breaking change. I don't think this class was useful for users (as a `public` class) and I don't think they used it. Nevertheless, we already had surprises with such changes.

I looked at the other types in the `Cloud` folder and I see some others which are `public` as well (`CloudMetadataProviderCollection` and `ICloudMetadataProvider`) - from what I see there is no way users can do anything useful with those. I think when we introduced cloud detection, we did not really think about the visibility of these types.

I don't have strong opinion on this specific change, only thing I'd add is:
- If we make this technically breaking change, then probably it'd make sense to think about making those other types `internal` as well.
- Or we just leave it `public`, since actually that's not really the point of this PR - and we may or may not clean up this later. With this we are definitely on the safe side.

Reply:
While I also think that the types in `Elastic.Apm.Cloud` should be `internal` rather than `public`, I agree that this is not the right PR to make this change. I will revert `AzureAppServiceMetadataProvider` to `public` visibility.

I will leave the newly introduced type `AzureFunctionsMetadataProvider` `internal` though, if that's ok with you.

I'll create an issue for that with the `breaking change` label. Then we can make these changes at a later point if we decide to do so.

Reply:
> Update: While changing back the visibility, I remembered the reason for this change. Leaving `AzureAppServiceMetadataProvider` `public` requires me to make the newly introduced base class `EnvironmentBasedAzureMetadataProvider` `public` as well, which I wanted to avoid. I suppose, I have to create a bit more legacy here.

Oh, I see. Yeah I think the current status of the PR is fine - with keeping the `AzureAppServiceMetadataProvider` public and accordingly making the new base public public as well is the safest way to proceed (definitely not breaking users). https://github.com/elastic/apm-agent-dotnet/issues/1890 is useful, let's handle it there...